system_metrics_collector/include/moving_average_statistics/moving_average.hpp

```diff
@@ -0,0 +1,189 @@
+   *  Returns the standard deviation (population) of all data recorded. If size of list is zero, returns NaN.
+   *
+   *  Variance is obtained by Welford's online algorithm,
+   *  see https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford%27s_online_algorithm
+   *
+   *  @return The standard deviation (population) of all data recorded, or NaN if size of data is zero.
+  **/
+  double standardDeviation()
+  {
+    return getStatistics().standard_deviation;
+  }
+  /**
+   *  Return a StatisticResults object, containing average, minimum, maximum, standard deviation (population),
+   *  and sample count.
+   *  For the case of no observations, the average, min, max, and standard deviation are NaN.
+   *
+   *  @return StatisticResults object, containing average, minimum, maximum, standard deviation (population),
+   *  and sample count.
+  **/
+  StatisticResults getStatistics()
+  {
+    std::lock_guard<std::recursive_mutex> guard(mutex);
+    StatisticResults to_return;
+    to_return.sample_count = count_;
+
+    if (count_ == 0) {
+      to_return.average = nan("");
+      to_return.min = nan("");
+      to_return.max = nan("");
+      to_return.standard_deviation = nan("");
+    } else {
+      to_return.average = average_;
+      to_return.min = min_;
+      to_return.max = max_;
+      to_return.standard_deviation = std::sqrt(sum_of_square_diff_from_mean_ / count_);
+    }
+
+    return to_return;
+  }
+  /**
+   *  Reset all calculated values. Equivalent to a new window for a moving average.
+  **/
+  void reset()
+  {
+    std::lock_guard<std::recursive_mutex> guard(mutex);
+    average_ = 0;
+    min_ = DBL_MAX;
+    max_ = DBL_MIN;
+    sum_of_square_diff_from_mean_ = 0;
+    count_ = 0;
+  }
+  /**
+   *  Observe a sample for the given window. The input item is used to calculate statistics.
+   *
+   *  Derived class should override add_item() and call add_item_() with custom input.
+   *
+   *  @param item The item that was observed
+  **/
+  virtual void add_item(const T & item)
+  {
+    add_item_(item);
+  }
+
+  /**
+   * Return the number of samples observed
+   *
+   * @return the number of samples observed
+   */
+  virtual int64_t get_count()
+  {
+    std::lock_guard<std::recursive_mutex> guard(mutex);
+    return count_;
+  }
+
+protected:
+  /**
+   *  Internal function for marking an observation. It handles data insertion and cache data validation logic.
+   *
+   *  Variance is obtained by Welford's online algorithm,
+   *  see https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford%27s_online_algorithm
+   *
+   *  @param item item to be stored.
+   */
+  void add_item_(const T & item)
+  {
+    std::lock_guard<std::recursive_mutex> guard(mutex);
+    count_++;
+    double previous_average_ = average_;
+    average_ = previous_average_ + (item - previous_average_) / count_;
+    min_ = std::min(min_, static_cast<double>(item));
+    max_ = std::max(max_, static_cast<double>(item));
```

Comment:
Let's drop the template if you're anyway going to cast everything back to double here, this presents little advantage.